nengo/builder/optimizer.py

```diff
@@ -0,0 +1,934 @@
+from nengo.utils.stdlib import Timer, WeakKeyDefaultDict, WeakSet
+
+logger = logging.getLogger(__name__)
+
+__all__ = ["optimize"]
+
+
+def optimize(model, dg, max_passes=None):
+    """Optimizes the operator graph by merging operators.
+
+    This reduces the number of iterators to iterate over in slow Python code
+    (as opposed to fast C code). The resulting merged operators will also
+    operate on larger chunks of sequential memory, making better use of CPU
+    caching and prefetching.
+
+    The optimization algorithm has worst case complexity :math:`O(n^2 + e)`,
+    where :math:`n` is the number of operators and :math:`e` is the number
+    of edges in the dependency graph. In practice the run time will be much
+    better because not all :math:`n^2` pairwise combinations of operators
+    will be evaluated. A grouping depending on the operator type and view
+    bases is done with dictionaries. This grouping can be done in amortized
+    linear time and reduces the actual worst-case runtime of the optimization
+    algorithm to :math:`O(gm^2 + e)`, where :math:`g` is the number of groups
+    and :math:`m` is the number of elements in a group. Moreover, information
+    about memory alignment will be used to cut the inner loop short in
+    many cases and gives a runtime much closer to linear in most cases.
+
+    Note that this function modifies both ``model`` and ``dg``.
+
+    Parameters
+    ----------
+    model : `nengo.builder.Model`
+        Builder output to optimize.
+    dg : dict
+        Dict of the form ``{a: {b, c}}`` where ``b`` and ``c`` depend on ``a``,
+        specifying the operator dependency graph of the model.
+    """
+
+    logger.info("Optimizing model...")
+
+    # We try first to merge operators with views only as these have a fixed
+    # order for the memory alignment whereas operators without views could
+    # be merged in a random order. Merging the views of operators will
+    # propagate requirements in the memory ordering via the other
+    # associated signals of the operator to other operators.
+
+    # Once no more operators with views can be merged, we try to merge
+    # operators without views and then try again merging views (because
+    # each operator merge might generate new views).
+
+    single_pass = OpMergePass(dg)
+
+    n_initial_ops = len(dg)
+    cum_duration = 0.
+    before, after = None, None
+    i = 0
+    only_merge_ops_with_view = True
+    while only_merge_ops_with_view or after < before:
+        if max_passes is not None and i >= max_passes:
+            break
+
+        i += 1
+
+        if logger.isEnabledFor(logging.DEBUG):
+            for tp, ops in iteritems(groupby(dg, type)):
+                logger.debug("%s: %d", tp, len(ops))
+
+        only_merge_ops_with_view = before is None or before != after
+        before = len(dg)
+
+        with Timer() as t:
+            single_pass(only_merge_ops_with_view)
+
+        after = len(dg)
```

Comment:
> Is this a weird expectation to have?

I don't think those are pretty reasonable expectations, but unfortunately things are a bit more complicated here.

> if the call would return and over-write the dg instance, would it affect performance?

Yes, it would increase memory usage; potentially by a lot for large models. The dependency graph can get quite large itself already, but then it also references the initial set of operators and those reference the initial set of signals that require even more memory with their initial values. By keeping the old dependency graph around all those objects would be prevented from being freed.

Maybe it would be clearer to use `single_pass.dg.forward` instead of `dg` to make it clear that it is an attribute of the `single_pass` instance that might get modified by the object?

> I have a mental model of Classes as containers for data

I think that is an appropriate mental model in many cases, but not all. Especially here `OpMergePass` is more like a [function object (sometimes called functor)](https://en.wikipedia.org/wiki/Function_object). It is used pretty much like a function, but by using an object it allows us to split up the algorithm into smaller functions, but still group them. It also allows to easily share data between those functions and preserve some state across invocations.